autonat.go

```diff
@@ -0,0 +1,178 @@
+	NATStatusPrivate
+)
+
+var (
+	AutoNATBootDelay       = 15 * time.Second
+	AutoNATRefreshInterval = 15 * time.Minute
+	AutoNATRequestTimeout  = 60 * time.Second
+)
+
+// AutoNAT is the interface for ambient NAT autodiscovery
+type AutoNAT interface {
+	// Status returns the current NAT status
+	Status() NATStatus
+	// PublicAddr returns the public dial address when NAT status is public and an
+	// error otherwise
+	PublicAddr() (ma.Multiaddr, error)
+}
+
+// AmbientAutoNAT is the implementation of ambient NAT autodiscovery
+type AmbientAutoNAT struct {
+	ctx  context.Context
+	host host.Host
+
+	mx     sync.Mutex
+	peers  map[peer.ID]struct{}
+	status NATStatus
+	addr   ma.Multiaddr
+}
+
+// NewAutoNAT creates a new ambient NAT autodiscovery instance attached to a host
+func NewAutoNAT(ctx context.Context, h host.Host) AutoNAT {
+	as := &AmbientAutoNAT{
+		ctx:    ctx,
+		host:   h,
+		peers:  make(map[peer.ID]struct{}),
+		status: NATStatusUnknown,
+	}
+
+	h.Network().Notify(as)
+	go as.background()
+
+	return as
+}
+
+func (as *AmbientAutoNAT) Status() NATStatus {
+	return as.status
+}
+
+func (as *AmbientAutoNAT) PublicAddr() (ma.Multiaddr, error) {
+	as.mx.Lock()
+	defer as.mx.Unlock()
+
+	if as.status != NATStatusPublic {
+		return nil, errors.New("NAT Status is not public")
+	}
+
+	return as.addr, nil
+}
+
+func (as *AmbientAutoNAT) background() {
+	// wait a bit for the node to come online and establish some connections
+	// before starting autodetection
+	select {
+	case <-time.After(AutoNATBootDelay):
+	case <-as.ctx.Done():
+		return
+	}
+
+	for {
+		as.autodetect()
+		select {
+		case <-time.After(AutoNATRefreshInterval):
+		case <-as.ctx.Done():
+			return
+		}
+	}
+}
+
+func (as *AmbientAutoNAT) autodetect() {
+	peers := as.getPeers()
+
+	if len(peers) == 0 {
+		log.Debugf("skipping NAT auto detection; no autonat peers")
+		return
+	}
+
+	cli := NewAutoNATClient(as.host)
+
+	for _, p := range peers {
+		ctx, cancel := context.WithTimeout(as.ctx, AutoNATRequestTimeout)
+		a, err := cli.DialBack(ctx, p)
+		cancel()
+
+		switch {
+		case err == nil:
+			log.Debugf("NAT status is public; address through %s: %s", p.Pretty(), a.String())
+			as.mx.Lock()
+			as.addr = a
+			as.status = NATStatusPublic
+			as.mx.Unlock()
+			return
+
+		case IsDialError(err):
+			log.Debugf("NAT status is private; dial error through %s: %s", p.Pretty(), err.Error())
+			as.mx.Lock()
+			as.status = NATStatusPrivate
+			as.mx.Unlock()
+			return
+
+		default:
+			log.Debugf("Error dialing through %s: %s", p.Pretty(), err.Error())
+		}
+	}
+
+	as.mx.Lock()
+	as.status = NATStatusUnknown
+	as.mx.Unlock()
+}
+
+func (as *AmbientAutoNAT) getPeers() []peer.ID {
+	as.mx.Lock()
+	defer as.mx.Unlock()
+
+	if len(as.peers) == 0 {
+		return nil
+	}
+
+	peers := make([]peer.ID, 0, len(as.peers))
+	for p := range as.peers {
+		if as.host.Network().Connectedness(p) == inet.Connected {
+			peers = append(peers, p)
+		}
+	}
+
+	if len(peers) == 0 {
```

Comment:
I'm afraid this black or white decision could yield erratic results, e.g. if you only have 1 active connection to an autonat peer, we're going to restrict our query to a single peer. I much rather have a minimum threshold we strive for, e.g. 5 peers, for resilience purposes, starting with peers we hold a connection to. 

As it is, the `autodetect` does a round-robin, so no risk of establishing redundant connections if we shuffle the connected and non-connected sublists separately, i.e. 

```
A..G: connected
U..Z: not connected

A B C D E F G || U V W X Y Z
<- shuffle ->   <- shuffle ->
```

Reply:
> The code tries to use an already existing connection purely to avoid creating unnecessary new connections.

Currently, if we happen to be connected to 1 autonat peer only, we'll restrict ourselves to it. If it fails, we're out of luck. This makes us fragile, especially because we expect scarcity in autonat peers.

What I'm proposing is to target N peers (e.g. 5), preferring connected peers, and falling back to disconnected ones to fill up the slice. To avoid connected and unconnected peers getting mixed up in the shuffle, we keep track of the pivot index and shuffle both sublists separately.

Since `autodetect` is round-robin, it'll only resort to disconnected peers if the connected ones fail. This makes us more resilient overall.

Reply:
That's fine, but do we want to dial more than one peers when we get a DIAL_ERROR?

Reply:
That's a good question. I'm not sure I have the answer. Right now we flag `NATStatusPrivate` on the first dial err, and abort. However, what if that peer is behind some kind of firewall (corporate, geographical, etc.)? Wouldn't it be better to get more observations?

Reply:
> I guess we could do a few more tries if we have more known autonat peers, but accept the failure if we don't have enough.

Yep. If we don't have enough, we'd defer to the next iteration. If by then we've found more autonat peers, with this new logic we'll query them even if not connected, and hence have a chance to improve our connectivity.

> I would arbitrarily go for "3 times is enemy action".

We detect "enemy action" on the receiving side through the throttling, no? (3 is fine for that) 

That makes me realise that we should probably move peers who have sent us `E_DIAL_REFUSED` or `E_INTERNAL_ERROR` to a blacklist, to avoid dwelling on them, and to be well-behaved.

Reply:
That's probably too much complexity for marginal improvement :)

Also, I think I want some slightly more clever strategy for making multiple dial attempts -- if our nat status was unknown or public, then try 3 times. 
If it was private, then a single failure should be enough to convince us.

Reply:
Implemented the "3 times is enemy action" strategy in https://github.com/libp2p/go-libp2p-autonat/pull/1/commits/aadb8dbcb5d0a8a3a006f546255b4cad871b6d94, with memory of past failures so that it stops asking multiple peers once it has enough confidence we are NATed.

Reply:
In https://github.com/libp2p/go-libp2p-autonat/pull/1/commits/d7f55b0524da51340ab00fc38f4b06b34b87f01c we ensure that we have at least 3 autonat peers in the candidate set, even when we are connected to less than that.
This uses the strategy you suggested for ordering.